File: DID.Java.SDK/src/main/java/org/elastos/did/Proof.java
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.did;

import java.io.IOException;
import java.util.Calendar;
import java.util.Date;

import org.elastos.did.util.JsonHelper;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonNode;

public class Proof {
	private String type;
	private Date created;
	private DIDURL creator;
	private String signature;

	protected Proof(String type, Date created, DIDURL creator, String signature) {
		this.type = type;
		this.created = created;
		this.creator = creator;
		this.signature = signature;
	}

	protected Proof(DIDURL creator, String signature) {
		this(Constants.defaultPublicKeyType,
				Calendar.getInstance(Constants.UTC).getTime(),
				creator, signature);
	}

    public String getType() {
    	return type;
    }

    public Date getCreated() {
    	return created;
    }

    public DIDURL getCreator() {
    	return creator;
    }

    public String getSignature() {
    	return signature;
    }

	protected static Proof fromJson(JsonNode node, DIDURL refSignKey)
			throws MalformedDocumentException {
		Class<MalformedDocumentException> clazz = MalformedDocumentException.class;

		String type = JsonHelper.getString(node, Constants.type,
				true, Constants.defaultPublicKeyType,
				"document proof type", clazz);

		Date created = JsonHelper.getDate(node, Constants.created,
				true, null, "proof created date", clazz);

		DIDURL creator = JsonHelper.getDidUrl(node, Constants.creator, true,
				refSignKey.getDid(), "document proof creator", clazz);
		if (creator == null)
			creator = refSignKey;

		String signature = JsonHelper.getString(node, Constants.signatureValue,
				false, null, "document proof signature", clazz);

		return new Proof(type, created, creator, signature);
	}

	protected void toJson(JsonGenerator generator, boolean normalized)
			throws IOException {
		generator.writeStartObject();

		// type
		if (normalized || !type.equals(Constants.defaultPublicKeyType)) {
			generator.writeFieldName(Constants.type);
			generator.writeString(type);
		}

		// created
		if (created != null) {
			generator.writeFieldName(Constants.created);
			generator.writeString(JsonHelper.format(created));
		}

		// creator
		if (normalized) {
			generator.writeFieldName(Constants.creator);
			generator.writeString(creator.toString());
		}

		// signature
		generator.writeFieldName(Constants.signatureValue);
		generator.writeString(signature);

		generator.writeEndObject();
	}
}

File: DID.Swift.SDK/ElastosDIDSDK/DID/Proof.swift
import Foundation

public class Proof {
    public var type: String!
    public var created: Date?
    public var creator: DIDURL?
    public var signature: String!
    public var verificationMethod: DIDURL!
    
    public var realm: String?
    public var nonce: String?

    // init for document
    init(_ type: String, _ created: Date, _ creator: DIDURL, _ signature: String) {
        self.type = type
        self.created = created
        self.creator = creator
        self.signature = signature
    }
    
    // init for document
    init(_ creator: DIDURL, _ signature: String) {
        self.type = DEFAULT_PUBLICKEY_TYPE
        self.created = DateFormater.currentDate()
        self.creator = creator
        self.signature = signature
    }
    
    // init for VerifiableCredential
    init(_ type: String, _ method: DIDURL, _ signature: String) {
        self.type = type
        self.verificationMethod = method
        self.signature = signature
    }

    init(_ type: String, _ method: DIDURL, _ realm: String, _ nonce: String, _ signature: String) {
        self.type = type
        self.verificationMethod = method
        self.realm = realm
        self.nonce = nonce
        self.signature = signature
    }
    
    init(_ method: DIDURL, _ realm: String, _ nonce: String, _ signature: String) {
        self.type = DEFAULT_PUBLICKEY_TYPE
        self.verificationMethod = method
        self.realm = realm
        self.nonce = nonce
        self.signature = signature
    }
    
    func toJson_vc(_ ref: DID, _ normalized: Bool) -> OrderedDictionary<String, Any> {
        var dic: OrderedDictionary<String, Any> = OrderedDictionary()
        var value: String
        //type:
        if normalized || type != DEFAULT_PUBLICKEY_TYPE {
            dic[TYPE] = type
        }
        
        // method:
        if normalized || verificationMethod.did != ref {
             value = verificationMethod.toExternalForm()
        }
        else {
            value = "#" + verificationMethod.fragment
        }
        dic[VERIFICATION_METHOD] = value
        
        // signature:
        dic[SIGNATURE] = signature
        return dic
    }

    func toJson_dc(_ normalized: Bool) -> OrderedDictionary<String, Any> {
        var dic: OrderedDictionary<String, Any> = OrderedDictionary()
        //type:
        if normalized || (type != DEFAULT_PUBLICKEY_TYPE) {
            dic[TYPE] = type
        }
        
        // created
        if created != nil {
            dic[CREATED] = DateFormater.format(created!)
        }
        
        // creator
        if (normalized) {
            dic[CREATOR] = creator!.toExternalForm()
        }
        // signature:
        dic[SIGNATURE_VALUE] = signature
        
        return dic
    }

    func toJson_vp() -> OrderedDictionary<String, Any> {
        var dic: OrderedDictionary<String, Any> = OrderedDictionary()
        var value: String
        //type:
        dic[TYPE] = type
        
        // method:
        value = verificationMethod.toExternalForm()
        dic[VERIFICATION_METHOD] = value
        
        // realm
        dic[REALM] = realm!
        
        // nonce
        dic[NONCE] = nonce!
        
        // signature:
        dic[SIGNATURE] = signature

        return dic
    }
    
    class func fromJson_dc(_ json: OrderedDictionary<String, Any>, _ refSignKey: DIDURL) throws -> Proof {
        let type: String = try JsonHelper.getString(json, TYPE, true, DEFAULT_PUBLICKEY_TYPE, "document proof type")
        
        let created: Date = try DateFormater.getDate(json, CREATED, true, nil, "")!
        
        let creator = try JsonHelper.getDidUrl(json, CREATOR, true, refSignKey.did, "document proof creator")
        var c = creator
        if creator == nil {
            c = refSignKey
        }
        let signature: String = try JsonHelper.getString(json, SIGNATURE_VALUE, false, nil, "document proof signature")
        return Proof(type, created, c!, signature)
    }

    // for VerifiableCredential
    class func fromJson_vc(_ json: OrderedDictionary<String, Any>, _ ref: DID?) throws -> Proof {
        let type: String = try JsonHelper.getString(json, TYPE, true, DEFAULT_PUBLICKEY_TYPE, "crendential proof type")
        let method: DIDURL = try JsonHelper.getDidUrl(json, VERIFICATION_METHOD, ref, "crendential proof verificationMethod")!
        let signature: String = try JsonHelper.getString(json, SIGNATURE, false, nil, "crendential proof signature")
        return Proof(type, method, signature)
    }

    class func fromJson_vp(_ json: OrderedDictionary<String, Any>, _ ref: DID?) throws -> Proof {
        let type: String = try JsonHelper.getString(json, TYPE, true, DEFAULT_PUBLICKEY_TYPE, "crendential proof type")
        let method: DIDURL = try JsonHelper.getDidUrl(json, VERIFICATION_METHOD, ref, "presentation proof verificationMethod")!
         let realm: String = try JsonHelper.getString(json, REALM, false, nil, "presentation proof realm")
         let nonce: String = try JsonHelper.getString(json, NONCE, false, nil, "presentation proof nonce")
        let signature: String = try JsonHelper.getString(json, SIGNATURE, false, nil, "presentation proof signature")
        return Proof(type, method, realm, nonce, signature)
    }
}
